File: src/libkali/kali/core/tests/test_timer.cpp
#include "testing.h"
#include "kali/core/timer.h"

#include <cstdint>
#include <thread>
#include <chrono>

using namespace kali;

TEST_SUITE_BEGIN("timer");

TEST_CASE("now")
{
    Timer::TimePoint t0 = Timer::now();
    std::this_thread::sleep_for(std::chrono::milliseconds(10));
    Timer::TimePoint t1 = Timer::now();
    CHECK_GT(t1, t0 + 10000000 / 2);
}

TEST_CASE("delta")
{
    CHECK_EQ(Timer::delta_s(0, 1000000000), 1.0);
    CHECK_EQ(Timer::delta_ms(0, 1000000000), 1000.0);
    CHECK_EQ(Timer::delta_us(0, 1000000000), 1000000.0);
    CHECK_EQ(Timer::delta_ns(0, 1000000000), 1000000000.0);
}

TEST_SUITE_END();

File: src/libkali/kali/core/tests/test_timer.py
import pytest
from kali import Timer
import time


def test_now():
    delta_ns = 10000000
    t0 = Timer.now()
    time.sleep(delta_ns / 1000000000.0)
    t1 = Timer.now()
    assert t1 > t0 + delta_ns / 2


def test_delta():
    assert Timer.delta_ns(0, 1000000000) == 1000000000.0
    assert Timer.delta_us(0, 1000000000) == 1000000.0
    assert Timer.delta_ms(0, 1000000000) == 1000.0
    assert Timer.delta_s(0, 1000000000) == 1.0


def test_timer():
    delta_ns = 10000000
    delta_us = delta_ns / 1000.0
    delta_ms = delta_us / 1000.0
    delta_s = delta_ms / 1000.0
    timer = Timer()
    time.sleep(delta_ns / 1000000000.0)
    assert timer.elapsed_ns() > delta_ns / 2
    assert timer.elapsed_us() > delta_us / 2
    assert timer.elapsed_ms() > delta_ms / 2
    assert timer.elapsed_s() > delta_s / 2


if __name__ == "__main__":
    pytest.main([__file__])
